### PR TITLE
copr: support timezone (fixed offset)

### PR DESCRIPTION
**Pull request by BusyJay**
We need to compare string with timestamp sometimes, it will get a wrong result if not taking timezone into consideration.

@ngaut @coocood @shenli @qiuyesuifeng PTAL

### CODE REVIEW

File: src/util/codec/mysql/time.rs [hunk 1]
```diff
@@ -230,18 +235,25 @@ impl Time {
...
+                                   d,
+                                   h,
+                                   minute,
+                                   sec,
+                                   frac * TEN_POW[9 - fsp as usize]));
         Time::new(t, types::DATETIME as u8, fsp as i8)
     }
 
-    pub fn from_packed_u64(t: u64, tp: u8, fsp: i8) -> Result<Time> {
+    pub fn from_packed_u64(t: u64, tp: u8, fsp: i8, tz: &FixedOffset) -> Result<Time> {
```
Comment (coocood) [inline src/util/codec/mysql/time.rs]: Add comments about `tz` is not used for TIMESTAMP.

File: src/util/codec/mysql/time.rs [hunk 2]
```diff
@@ -255,15 +267,24 @@ impl Time {
...
+        let t = if tp == types::TIMESTAMP {
+            let t = try!(ymd_hms_nanos(&UTC, year, month, day, hour, minute, second, nanosec));
+            tz.from_utc_datetime(&t.naive_utc())
+        } else {
+            try!(ymd_hms_nanos(tz, year, month, day, hour, minute, second, nanosec))
+        };
         Time::new(t, tp, fsp as i8)
     }
 
     pub fn to_packed_u64(&self) -> u64 {
```
Comment (coocood) [inline src/util/codec/mysql/time.rs]: Add comments about difference for Timestamp and other types.

File: src/util/codec/mysql/time.rs (hunk 1 above)
Comment (BusyJay) [inline src/util/codec/mysql/time.rs]: No, `tz` is used for TIMESTAMP too, see L272.

File: src/util/codec/mysql/time.rs [hunk 3]
```diff
@@ -404,10 +437,22 @@ mod test {
...
+                offset *= 60;
+                let tz = FixedOffset::east(offset);
+                let aware_t = Time::parse_datetime(s, fsp, &tz).unwrap();
+                let aware_packed = aware_t.to_packed_u64();
+                let reverted = Time::from_packed_u64(aware_packed, types::DATETIME, fsp, &tz)
+                    .unwrap();
+                assert_eq!(reverted, aware_t);
+                assert_eq!(reverted.to_packed_u64(), aware_packed);
+
+                let revered_timestap =
```
Comment (coocood) [inline src/util/codec/mysql/time.rs]: reverted_timestamp

File: src/util/codec/mysql/time.rs [hunk 4]
```diff
@@ -404,10 +437,22 @@ mod test {
...
-            let t = Time::parse_datetime(s, fsp).unwrap();
-            let packed = t.to_packed_u64();
-            let reverted = Time::from_packed_u64(packed, types::DATETIME, fsp).unwrap();
-            assert_eq!(reverted, t);
+            for mut offset in MIN_OFFSET..MAX_OFFSET {
+                offset *= 60;
+                let tz = FixedOffset::east(offset);
+                let aware_t = Time::parse_datetime(s, fsp, &tz).unwrap();
+                let aware_packed = aware_t.to_packed_u64();
+                let reverted = Time::from_packed_u64(aware_packed, types::DATETIME, fsp, &tz)
```
Comment (coocood) [inline src/util/codec/mysql/time.rs]: reverted_datetime

File: src/util/codec/mysql/time.rs [hunk 5]
```diff
@@ -404,10 +437,22 @@ mod test {
...
         for (s, fsp) in cases {
-            let t = Time::parse_datetime(s, fsp).unwrap();
-            let packed = t.to_packed_u64();
-            let reverted = Time::from_packed_u64(packed, types::DATETIME, fsp).unwrap();
-            assert_eq!(reverted, t);
+            for mut offset in MIN_OFFSET..MAX_OFFSET {
+                offset *= 60;
+                let tz = FixedOffset::east(offset);
+                let aware_t = Time::parse_datetime(s, fsp, &tz).unwrap();
+                let aware_packed = aware_t.to_packed_u64();
```
Comment (coocood) [inline src/util/codec/mysql/time.rs]: Now we use timezone aware DateTime, we don't need `aware` at variable name.

File: src/util/codec/mysql/time.rs [hunk 6]
```diff
@@ -230,18 +235,28 @@ impl Time {
...
         let frac = try!(parse_frac(frac_str.as_bytes(), fsp));
         if y == 0 && m == 0 && d == 0 && h == 0 && minute == 0 && sec == 0 {
-            return Ok(zero_datetime());
+            return Ok(zero_datetime(tz));
         }
         if y < 0 || y > 9999 {
             return Err(box_err!("unsupport year: {}", y));
         }
-        let t = try!(ymd_hms_nanos(y, m, d, h, minute, sec, frac * TEN_POW[9 - fsp as usize]));
+        let t = try!(ymd_hms_nanos(tz,
```
Comment (coocood) [inline src/util/codec/mysql/time.rs]: NaiveDateTime type should have a different name than 't', so we can infer the type by its name.

File: src/util/codec/mysql/time.rs [hunk 7]
```diff
@@ -230,18 +235,28 @@ impl Time {
...
+                                   sec,
+                                   frac * TEN_POW[9 - fsp as usize]));
         Time::new(t, types::DATETIME as u8, fsp as i8)
     }
 
-    pub fn from_packed_u64(t: u64, tp: u8, fsp: i8) -> Result<Time> {
+    /// Get time from packed u64. When `tp` is `TIMESTAMP`, the packed time should
+    /// be a UTC time; otherwise the packed time should be in the same timezone as `tz`
+    /// specified.
+    pub fn from_packed_u64(t: u64, tp: u8, fsp: i8, tz: &FixedOffset) -> Result<Time> {
```
Comment (coocood) [inline src/util/codec/mysql/time.rs]: In this function, type of 't' has changed from u64 to NaiveDateTime, it's hard to read.
We should have different name pattern for packed u64 time, timezone aware datetime and naive datetime

File: src/util/codec/mysql/time.rs (hunk 6 above)
Comment (BusyJay) [inline src/util/codec/mysql/time.rs]: No, it's not a `NaiveDateTime`. And I don't think add any type related information to name is a good idea.

**Comment by coocood** [comment]
LGTM

**Comment by BusyJay** [comment]
@zhangjinpeng1987 PTAL

**Comment by BusyJay** [comment]
@hhkbp2 PTAL

File: src/util/xeval/evaluator.rs [hunk 8]
```diff
@@ -24,6 +24,33 @@ use std::collections::HashMap;
 use std::cmp::Ordering;
 use std::ascii::AsciiExt;
 use tipb::expression::{Expr, ExprType};
+use tipb::select::SelectRequest;
+use chrono::FixedOffset;
+
+#[derive(Debug)]
+pub struct EvalContext {
```
Comment (hhkbp2) [inline src/util/xeval/evaluator.rs]: Could use some comment for members of this context.

File: src/util/xeval/evaluator.rs [hunk 9]
```diff
@@ -24,6 +24,33 @@ use std::collections::HashMap;
...
+impl Default for EvalContext {
+    fn default() -> EvalContext {
+        EvalContext { tz: FixedOffset::east(0) }
+    }
+}
+
+impl EvalContext {
+    pub fn new(sel: &SelectRequest) -> Result<EvalContext> {
+        let offset = sel.get_time_zone_offset();
+        if offset <= -3600 * 24 || offset >= 3600 * 24 {
```
Comment (hhkbp2) [inline src/util/xeval/evaluator.rs]: Make a constant for 3600 * 24.

**Comment by hhkbp2** [comment]
Rest LGTM

**Comment by hhkbp2** [comment]
LGTM

**Pull request merged**